{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7c473a8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import h5py\n",
    "import gzip\n",
    "from Bio import SeqIO\n",
    "import os\n",
    "from mimetypes import guess_type\n",
    "from functools import partial\n",
    "import time\n",
    "import json\n",
    "import numpy as np\n",
    "from refGenome import refGenome_h5\n",
    "import seaborn as sb\n",
    "import matplotlib.pyplot as plt\n",
    "from Dataset import SubIntervalDataset\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "54b32562-f8ac-44aa-b895-6e4bb5a0d9ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "#@numba.jit(parallel=True)\n",
    "def build_h5(fastas_path, \n",
    "             output_file_name, \n",
    "             encoding_dict={'A':0, 'C':1, 'T':2, 'G':3, 'N':4, 'M':4, 'R':4}, \n",
    "             files_prefix=\"Homo_sapiens.GRCh38.dna.chromosome.\",\n",
    "             chromosomes_list=list(range(1,23)) + ['X','Y','MT'],\n",
    "             file_extension = 'fa.gz',\n",
    "             chunk_size=(100000,)):\n",
    "    \"\"\"\n",
    "    https://stackoverflow.com/questions/42757283/seqio-parse-on-a-fasta-gz\n",
    "    \n",
    "    It will always read the first record only\n",
    "    Save the encoding dict as metadata in the h5 file\n",
    "    Each chromosome is a dataset\n",
    "    \n",
    "    @param files_prefix: the fasta files start with this string, followed by the number\\name of the chromosome\n",
    "    @param chromosomes: a list of the names that come after the prefix to use. You can use this to filter out \n",
    "                  chromosomes if not needed.\n",
    "                  \n",
    "    @param chunk_size: the chunk size of the h5 dataset. Data elements will be loaded in blocks of this size, so\n",
    "                       the larger the number is, the more RAM it takes but less disk IO\n",
    "    \"\"\"\n",
    "    #TODO: MAKE SURE THAT THE ENCODING_DICT TRANSLATES TO INTEGERS\n",
    "    output_file = h5py.File(output_file_name + \".h5\", \"a\")\n",
    "    \n",
    "    output_file[\"./\"].attrs[\"available_chromosomes\"] = []\n",
    "    output_file[\"./\"].attrs['time created'] = time.time()\n",
    "    \n",
    "    #Dictionaries are not natively supported in h5, and need to be serialized. Using json.dumps for that\n",
    "    metadata_dict = {}\n",
    "    metadata_dict[\"encoding_dict\"] = json.dumps(encoding_dict)\n",
    "    metadata_dict[\"available_chromosomes\"] = []\n",
    "    metadata_dict['chr_lengths'] = {}\n",
    "    metadata_dict['time created'] = time.time()\n",
    "    \n",
    "    for chromosome in chromosomes_list:\n",
    "        fname = f\"{files_prefix}{chromosome}.{file_extension}\"\n",
    "        file = os.path.join(fastas_path, fname)\n",
    "        \n",
    "        encoding = guess_type(file)[1]  # uses file extension\n",
    "        _open = partial(gzip.open, mode='rt') if encoding == 'gzip' else open\n",
    "        print(f\"\\tencoding {encoding}\")\n",
    "        \n",
    "        with _open(file) as handle:\n",
    "            print(f\"\\t{file} was opened successfully, parsing\")\n",
    "            print(\"\\tFound FASTA records in file:\")\n",
    "            \n",
    "            for i, record in enumerate(SeqIO.parse(handle, \"fasta\")):\n",
    "                \n",
    "                if i > 0:\n",
    "                    print(f\"\\tFile {file} contains more than one record, skipping them ({record})\")\n",
    "                    continue\n",
    "                \n",
    "                print(\"\\tSequence ID:\", record.id)\n",
    "                sequence = list(str(record.seq))\n",
    "                \n",
    "                sequence = [encoding_dict[ str.upper(bp) \n",
    "                                         ] for bp in sequence]\n",
    "                \n",
    "                seq_len = len(sequence)\n",
    "                print(\"\\tSequence Length:\", seq_len)\n",
    "                \n",
    "                chromosome_arr  = np.array(sequence, dtype=np.int8)\n",
    "                \n",
    "                #Save each as an integer of 1 byte\n",
    "                cs = chunk_size if chromosome != \"MT\" else True\n",
    "                output_file.create_dataset(f\"chr_{chromosome}\", data=chromosome_arr, dtype='i1', chunks=cs)\n",
    "                metadata_dict[\"available_chromosomes\"].append(record.id)\n",
    "                metadata_dict['chr_lengths'][f\"chr_{chromosome}\"] = seq_len\n",
    "                \n",
    "    print(metadata_dict)\n",
    "    output_file[\"./\"].attrs[\"metadata\"] = json.dumps(metadata_dict)\n",
    "    output_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4bf6d30c-3011-46a7-895a-30a80ffdf0f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tencoding gzip\n",
      "\t/lustre06/project/6065672/otastet1/ref/ref_by_chr/16.fa.gz was opened successfully, parsing\n",
      "\tFound FASTA records in file:\n",
      "\tSequence ID: 16\n",
      "\tSequence Length: 90354753\n",
      "\tencoding gzip\n",
      "\t/lustre06/project/6065672/otastet1/ref/ref_by_chr/17.fa.gz was opened successfully, parsing\n",
      "\tFound FASTA records in file:\n",
      "\tSequence ID: 17\n",
      "\tSequence Length: 81195210\n",
      "\tencoding gzip\n",
      "\t/lustre06/project/6065672/otastet1/ref/ref_by_chr/18.fa.gz was opened successfully, parsing\n",
      "\tFound FASTA records in file:\n",
      "\tSequence ID: 18\n",
      "\tSequence Length: 78077248\n",
      "\tencoding gzip\n",
      "\t/lustre06/project/6065672/otastet1/ref/ref_by_chr/19.fa.gz was opened successfully, parsing\n",
      "\tFound FASTA records in file:\n",
      "\tSequence ID: 19\n",
      "\tSequence Length: 59128983\n",
      "\tencoding gzip\n",
      "\t/lustre06/project/6065672/otastet1/ref/ref_by_chr/20.fa.gz was opened successfully, parsing\n",
      "\tFound FASTA records in file:\n",
      "\tSequence ID: 20\n",
      "\tSequence Length: 63025520\n",
      "\tencoding gzip\n",
      "\t/lustre06/project/6065672/otastet1/ref/ref_by_chr/21.fa.gz was opened successfully, parsing\n",
      "\tFound FASTA records in file:\n",
      "\tSequence ID: 21\n",
      "\tSequence Length: 48129895\n",
      "\tencoding gzip\n",
      "\t/lustre06/project/6065672/otastet1/ref/ref_by_chr/22.fa.gz was opened successfully, parsing\n",
      "\tFound FASTA records in file:\n",
      "\tSequence ID: 22\n",
      "\tSequence Length: 51304566\n",
      "{'encoding_dict': '{\"A\": 0, \"C\": 1, \"T\": 2, \"G\": 3, \"N\": 4, \"M\": 4, \"R\": 4}', 'available_chromosomes': ['16', '17', '18', '19', '20', '21', '22'], 'chr_lengths': {'chr_16': 90354753, 'chr_17': 81195210, 'chr_18': 78077248, 'chr_19': 59128983, 'chr_20': 63025520, 'chr_21': 48129895, 'chr_22': 51304566}, 'time created': 1666373954.8057423}\n"
     ]
    }
   ],
   "source": [
    "ref_genome_path = \"/lustre06/project/6065672/otastet1/ref/ref_by_chr/\"\n",
    "build_h5(ref_genome_path,\n",
    "         files_prefix =\"\",\n",
    "         chromosomes_list=list(range(16,23)),\n",
    "         output_file_name='/lustre06/project/6065672/otastet1/ref/test_chipseq/genome_test')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "08db9f8d-0845-43ca-a8d9-9a149c909944",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] Unable to open file (unable to open file: name = '/lustre06/project/6065672/otastet1/ref/test_chipseq/h5/genome_test', errno = 2, error message = 'No such file or directory', flags = 0, o_flags = 0)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [5], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m grch37_ds \u001b[38;5;241m=\u001b[39m \u001b[43mrefGenome_h5\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m/lustre06/project/6065672/otastet1/ref/test_chipseq/h5/genome_test\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/lustre06/project/6065672/otastet1/projects/prdm9_chipseq/recode/build_dataset/refGenome.py:10\u001b[0m, in \u001b[0;36mrefGenome_h5.__init__\u001b[0;34m(self, h5_path)\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__init__\u001b[39m(\u001b[38;5;28mself\u001b[39m, h5_path):\n\u001b[1;32m      8\u001b[0m \n\u001b[1;32m      9\u001b[0m     \u001b[38;5;66;03m# read file\u001b[39;00m\n\u001b[0;32m---> 10\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mf \u001b[38;5;241m=\u001b[39m \u001b[43mh5py\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mFile\u001b[49m\u001b[43m(\u001b[49m\u001b[43mh5_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mr\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m     12\u001b[0m     \u001b[38;5;66;03m# Print available datasets\u001b[39;00m\n\u001b[1;32m     13\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m key \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mf\u001b[38;5;241m.\u001b[39mkeys():\n",
      "File \u001b[0;32m/lustre06/project/6065672/otastet1/jupyter_env/lib/python3.8/site-packages/h5py/_hl/files.py:507\u001b[0m, in \u001b[0;36mFile.__init__\u001b[0;34m(self, name, mode, driver, libver, userblock_size, swmr, rdcc_nslots, rdcc_nbytes, rdcc_w0, track_order, fs_strategy, fs_persist, fs_threshold, fs_page_size, page_buf_size, min_meta_keep, min_raw_keep, locking, **kwds)\u001b[0m\n\u001b[1;32m    502\u001b[0m     fapl \u001b[38;5;241m=\u001b[39m make_fapl(driver, libver, rdcc_nslots, rdcc_nbytes, rdcc_w0,\n\u001b[1;32m    503\u001b[0m                      locking, page_buf_size, min_meta_keep, min_raw_keep, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwds)\n\u001b[1;32m    504\u001b[0m     fcpl \u001b[38;5;241m=\u001b[39m make_fcpl(track_order\u001b[38;5;241m=\u001b[39mtrack_order, fs_strategy\u001b[38;5;241m=\u001b[39mfs_strategy,\n\u001b[1;32m    505\u001b[0m                      fs_persist\u001b[38;5;241m=\u001b[39mfs_persist, fs_threshold\u001b[38;5;241m=\u001b[39mfs_threshold,\n\u001b[1;32m    506\u001b[0m                      fs_page_size\u001b[38;5;241m=\u001b[39mfs_page_size)\n\u001b[0;32m--> 507\u001b[0m     fid \u001b[38;5;241m=\u001b[39m \u001b[43mmake_fid\u001b[49m\u001b[43m(\u001b[49m\u001b[43mname\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43muserblock_size\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfapl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfcpl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mswmr\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mswmr\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    509\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(libver, \u001b[38;5;28mtuple\u001b[39m):\n\u001b[1;32m    510\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_libver \u001b[38;5;241m=\u001b[39m libver\n",
      "File \u001b[0;32m/lustre06/project/6065672/otastet1/jupyter_env/lib/python3.8/site-packages/h5py/_hl/files.py:220\u001b[0m, in \u001b[0;36mmake_fid\u001b[0;34m(name, mode, userblock_size, fapl, fcpl, swmr)\u001b[0m\n\u001b[1;32m    218\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m swmr \u001b[38;5;129;01mand\u001b[39;00m swmr_support:\n\u001b[1;32m    219\u001b[0m         flags \u001b[38;5;241m|\u001b[39m\u001b[38;5;241m=\u001b[39m h5f\u001b[38;5;241m.\u001b[39mACC_SWMR_READ\n\u001b[0;32m--> 220\u001b[0m     fid \u001b[38;5;241m=\u001b[39m \u001b[43mh5f\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mopen\u001b[49m\u001b[43m(\u001b[49m\u001b[43mname\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mflags\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfapl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfapl\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    221\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m mode \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mr+\u001b[39m\u001b[38;5;124m'\u001b[39m:\n\u001b[1;32m    222\u001b[0m     fid \u001b[38;5;241m=\u001b[39m h5f\u001b[38;5;241m.\u001b[39mopen(name, h5f\u001b[38;5;241m.\u001b[39mACC_RDWR, fapl\u001b[38;5;241m=\u001b[39mfapl)\n",
      "File \u001b[0;32mh5py/_objects.pyx:54\u001b[0m, in \u001b[0;36mh5py._objects.with_phil.wrapper\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mh5py/_objects.pyx:55\u001b[0m, in \u001b[0;36mh5py._objects.with_phil.wrapper\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mh5py/h5f.pyx:106\u001b[0m, in \u001b[0;36mh5py.h5f.open\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] Unable to open file (unable to open file: name = '/lustre06/project/6065672/otastet1/ref/test_chipseq/h5/genome_test', errno = 2, error message = 'No such file or directory', flags = 0, o_flags = 0)"
     ]
    }
   ],
   "source": [
    "grch37_ds = refGenome_h5('/lustre06/project/6065672/otastet1/ref/test_chipseq/h5/genome_test')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e3a06efc-1893-4a03-a66d-c9a407c76f49",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 0 19]\n",
      " [ 1 12]\n",
      " [ 2 47]\n",
      " [ 3 22]]\n"
     ]
    }
   ],
   "source": [
    "all_chr = grch37_ds.get_seq_by_chr_and_indices('chr_21',int(48129895/2),int(48129895/2)+100)\n",
    "unique, counts = np.unique(all_chr, return_counts=True)\n",
    "print(np.asarray((unique, counts)).T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "dd8fc524-daea-4179-b79d-37b803a426d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1, 2, 1, 0], dtype=int8)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_chr = grch37_ds.get_seq_by_chr_and_midpoint('chr_21',int(48129895/2),15)\n",
    "all_chr[:4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "ceae5c54-32fe-4c99-8164-3d67d96b7522",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['C', 'T', 'C', 'A']"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "decode = grch37_ds.convert_num_to_bases(all_chr[:4])\n",
    "decode[:4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "5d0eae1e-0e43-4161-8d28-39a56cf5dff6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1, 2, 1, 0]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grch37_ds.convert_bases_to_num(decode[:4])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "20967c13-1cd6-4f7b-a2f1-af0a31064bc5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot: >"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA/AAAACQCAYAAACvQ4DJAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/av/WaAAAACXBIWXMAAAsTAAALEwEAmpwYAAAU+UlEQVR4nO3dfbRldX3f8ffnnntHEa22khCZGQOpk5ipMYJ0tKUQGrQO1sW4modCkvpQlpOuBUYTV1uMLSa4kqW1K2naRdNMkYKJQgmJzSwz4WEZUl1pITNRVB4kmUxUZjAiarQGIgx8+8fZQ4733nPPncvdd5+H92utvebsffb5nc+Pczn3fvfvt/dOVSFJkiRJksbbXNcBJEmSJEnSaBbwkiRJkiRNAAt4SZIkSZImgAW8JEmSJEkTwAJekiRJkqQJYAEvSZIkSdIEsICXJEmSJGkdJbk6yYNJ7hryfJL85yQHk3wqyRmradcCXpIkSZKk9XUNsHOF588HtjXLbuBXV9OoBbwkSZIkSeuoqj4KfGWFXXYB76++24HnJHneqHYt4CVJkiRJ2libgfsH1g8321Y031qcxmMPHaq230PaCCeccnbXETbMIw98rOsIkiRJmhALJ31Xus7QpvlNm5fUtI8/9sBP0p/6fsyeqtrTepa230CSJEmSpEnVm1s6cf1ov1h/KgX7EWDrwPqWZtuKnEIvSZIkSdIQc5lbsqyDvcDrmqvRvxz4WlV9YdSLHIGXJEmSJGmIhbnecb8myXXAucBJSQ4D7wQWAKrqvwH7gFcDB4GHgTeupl0LeEmSJEmShuitYcS9qi4a8XwBlxxvuxbwkiRJkiQNsdw58F2xgJckSZIkaYj5NUyhb4sFvCRJkiRJQ6xlCn1bLOAlSZIkSRrCAl6SJEmSpAkwPzc+ZfP4JJEkSZIkacw4Ai9JkiRJ0gSYqAI+yQuBXcDmZtMRYG9V3dtmMEmSJEmSujaf8bkK/YqHEpL8W+B6IMAfNUuA65Jc1n48SZIkSZK608vckqUro975YuDvV9W7q+o3muXdwI7muWUl2Z3kQJIDV73/uvXMK0mSJEnShhmnAn7UFPongFOAzy3a/rzmuWVV1R5gD8BjDx2qpxJQkiRJkqSujNMU+lEF/FuBjyT5U+D+ZtvzgRcAl7aYS5IkSZKkzk3MReyq6qYk301/yvzgRez2V9XjbYeTJEmSJKlLCyPPPN84I69CX1VPALdvQBZJkiRJksbKWkfgk+wEfgXoAVc115MbfP75wLXAc5p9LquqfSu1OT6HEiRJkiRJGjNruYhdkh5wJXA+sB24KMn2Rbv9O+CGqjoduBD4r6PatYCXJEmSJGmIBeaWLKuwAzhYVYeq6lH6t2fftWifAv5W8/jZwAOjGh05hV6SJEmSpFm1xin0m/mbC8EDHAZetmifnwNuSfJm4ETgFaMadQRekiRJkqQhemTJkmR3kgMDy+41NH0RcE1VbQFeDfx6svLRAkfgJUmSJEkaYmGZmrqq9gB7VnjZEWDrwPqWZtugi4GdTXv/N8nTgZOAB4c16gi8JEmSJElD9JhbsqzCfmBbktOSbKJ/kbq9i/b5PHAeQJLvBZ4OfGmlRh2BlyRJkiRpiB457tdU1dEklwI3079F3NVVdXeSK4ADVbUXeBvw35P8NP0L2r2hqmqldi3gJUmSJEkaYrkp9KvR3NN936Jtlw88vgc463jabL2AP+GUs9t+i7HyyAMf6zrChvGzlTTuZul7yu8oSZLasZYR+LY4Ai9JkiRJ0hAW8JIkSZIkTYAFC3hJkiRJksbfvAW8JEmSJEnjzyn0kiRJkiRNAKfQS5IkSZI0ARyBlyRJkiRpAjgCL0mSJEnSBJivrhP8DQt4SZIkSZKG6HUdYIAFvCRJkiRJQziFXpIkSZKkCdBzCr0kSZIkSeNvnIrmubW+MMkb1zOIJEmSJEnjZqGWLquRZGeS+5IcTHLZkH1+NMk9Se5O8sFRba65gAd+foWgu5McSHLgiSf+6im8hSRJkiRJ3enV0mWUJD3gSuB8YDtwUZLti/bZBrwdOKuq/h7w1lHtrjgbIMmnhj0FnDzsdVW1B9gDML9p8xidMSBJkiRJ0uqtcQr9DuBgVR0CSHI9sAu4Z2CfNwFXVtVXAarqwaea5WTgVcBXF20P8H9Wl1uSJEmSpMm02inzi2wG7h9YPwy8bNE+3w2Q5A/p363u56rqppUaHVXAfxh4ZlXdufiJJH8w4rWSJEmSJE205abMJ9kN7B7YtKeZiX485oFtwLnAFuCjSb6vqv5ypRcMVVUXr/Dcjx1nOEmSJEmSJsp8La3gB08bH+IIsHVgfUuzbdBh4I6qegz48yR/Qr+g3z+s0adyETtJkiRJkqbaQtWSZRX2A9uSnJZkE3AhsHfRPv+L/ug7SU6iP6X+0EqNWsBLkiRJkjREb5lllKo6ClwK3AzcC9xQVXcnuSLJBc1uNwNfTnIPcBvwr6vqyyu1O073pJckSZIkaaws1BNrel1V7QP2Ldp2+cDjAn6mWVbFAl6SJEmSpCHGqWgepyySJEmSJI2VHmu7j1wbLOAlSZIkSRpiIWubQt8GC3hJkiRJkoZwBF6SJEmSpAnQm5uhEfhHHvhY228xVk445eyuI2yYWftsZ8ks/RzPmln7/3bW+qvp5HeyNHn8/TNd5mepgJckSZIkaVL15pxCL0mSJEnS2JupKfSSJEmSJE2q+Z4FvCRJkiRJY2+u5xR6SZIkSZLG3pwj8JIkSZIkjb/5+fEp4Oe6DiBJkiRJ0ria69WSZTWS7ExyX5KDSS5bYb8fSlJJzhzVpiPwkiRJkiQNMbeGEfgkPeBK4JXAYWB/kr1Vdc+i/Z4FvAW4Y1VZjjuJJEmSJEkzordQS5ZV2AEcrKpDVfUocD2wa5n93gW8B/jr1TRqAS9JkiRJ0hBrnEK/Gbh/YP1ws+1JSc4AtlbV7642i1PoJUmSJEkaYm5h6RT6JLuB3QOb9lTVntW2mWQO+CXgDceTZWQBn+SF9I8U3FFV3xjYvrOqbjqeN5MkSZIkaZLMLVM1N8X6SgX7EWDrwPqWZtsxzwJeBPxBEoDvAPYmuaCqDgzNslLQJD8F/A7wZuCuJINz9n9xpddKkiRJkjTpMr90WYX9wLYkpyXZBFwI7D32ZFV9rapOqqpTq+pU4HZgxeIdRp8D/ybgpVX1WuBc4N8necuxfgx7UZLdSQ4kOXDV+68b8RaSJEmSJI2nuU1Ll1Gq6ihwKXAzcC9wQ1XdneSKJBesNcuoYwdzx6bNV9Vnk5wL3JjkO1mhgB+cTvDYQ4dWd5M8SZIkSZLGTOaHlr4rqqp9wL5F2y4fsu+5q2lz1Aj8F5O8ZKDRbwCvAU4Cvm81byBJkiRJ0qTKfJYsXRk1Av864OjghmYqwOuS/FprqSRJkiRJGgPZND53X1+xgK+qwys894frH0eSJEmSpDHS4Yj7Yt4HXpIkSZKkITI/ISPwkiRJkiTNsiz0uo7wJAt4SZIkSZKGmbeAlyRJkiRp7DmFXpIkSZKkSbAwPmXz+CSRJEmSJGnMxCn0kiRJkiRNAEfgJUmSJEmaAI7AS5IkSZI0AXoW8FPrkQc+1nUE6Snz53h6nXDK2V1H2FCz9LM8a5/tLJmln2NpWszad/LRR490HaFVWVjoOsKTxud6+JIkSZIkjZv53tJlFZLsTHJfkoNJLlvm+Z9Jck+STyX5SJLvHNWmBbwkSZIkScP0ekuXEZL0gCuB84HtwEVJti/a7RPAmVX1YuBG4D+MatcCXpIkSZKkYRYWli6j7QAOVtWhqnoUuB7YNbhDVd1WVQ83q7cDW0Y16jnwkiRJkiQNscb7wG8G7h9YPwy8bIX9LwZ+b1SjFvCSJEmSJA3TW1o2J9kN7B7YtKeq9qyl+SQ/AZwJ/MCofS3gJUmSJEkaZpkp802xvlLBfgTYOrC+pdn2LZK8AngH8ANV9c1RUSzgJUmSJEkaZpkR+FXYD2xLchr9wv1C4McGd0hyOvBrwM6qenA1jVrAS5IkSZI0zCquOr9YVR1NcilwM9ADrq6qu5NcARyoqr3Ae4FnAr+ZBODzVXXBSu1awEuSJEmSNMzCpjW9rKr2AfsWbbt84PErjrdNC3hJkiRJkoZZ2xT6VoxMkmQHUFW1v7nx/E7gM83RBEmSJEmSptekFPBJ3gmcD8wnuZX+fetuAy5LcnpV/cIGZJQkSZIkqRtrnELfhrkRz/8wcBZwDnAJ8NqqehfwKuCfD3tRkt1JDiQ5cNX7r1u3sJIkSZIkbaje/NKlI6Pe+WhVPQ48nOTPqurrAFX1SJInhr1o8J54jz10qNYtrSRJkiRJG2mMRuBHFfCPJnlGVT0MvPTYxiTPBoYW8JIkSZIkTYP0FrqO8KRRBfw5VfVNgKoaLNgXgNe3lkqSJEmSpHEwKRexO1a8L7P9IeChVhJJkiRJkjQuFp7WdYInjc+hBEmSJEmSxs2kjMBLkiRJkjTLMjc558BLkiRJkjS7Jugq9JIkSZIkza4Jugq9JEmSJEmza4zOgZ/rOoAkSZIkSeMq85uWLKt6XbIzyX1JDia5bJnnn5bkfzbP35Hk1FFtWsBLkiRJkjRMb2HpMkKSHnAlcD6wHbgoyfZFu10MfLWqXgD8MvCeUe1awEuSJEmSNER680uWVdgBHKyqQ1X1KHA9sGvRPruAa5vHNwLnJclKjVrAS5IkSZI0zMLTli6jbQbuH1g/3Gxbdp+qOgp8DXjuSo22fjb+wknfteIRhLYk2V1Ve7p47402S30F+zvNZqmv0E1/jz56ZCPf7kl+tu3zs90Ys9TfWeorzFZ/Z6mv4Heynrrlatoku4HdA5v2bMR/+2kegd89epepMUt9Bfs7zWaprzBb/Z2lvsJs9XeW+gqz1d9Z6ivMVn9nqa8wW/2dpb52qqr2VNWZA8vi4v0IsHVgfUuzbdl9kswDzwa+vNL7TnMBL0mSJElSF/YD25KclmQTcCGwd9E+e4HXN49/GPj9qqqVGh2fG9pJkiRJkjQFqupokkuBm4EecHVV3Z3kCuBAVe0F3gf8epKDwFfoF/krmuYCfpbO/ZilvoL9nWaz1FeYrf7OUl9htvo7S32F2ervLPUVZqu/s9RXmK3+zlJfx15V7QP2Ldp2+cDjvwZ+5HjazIgRekmSJEmSNAY8B16SJEmSpAkwdQV8kp1J7ktyMMllXedpU5KrkzyY5K6us2yEJFuT3JbkniR3J3lL15nakuTpSf4oySebvv5815nalqSX5BNJPtx1lrYl+WySTye5M8mBrvO0LclzktyY5DNJ7k3yD7rO1IYk39N8pseWryd5a9e52pTkp5vvqLuSXJfk6V1nakuStzT9vHsaP9fl/qZI8neS3JrkT5t//3aXGdfLkL7+SPPZPpHkzC7zrbch/X1v8538qSQfSvKcDiOuqyH9fVfT1zuT3JLklC4zrpeVaoEkb0tSSU7qIpvaM1UFfJIecCVwPrAduCjJ9m5TteoaYGfXITbQUeBtVbUdeDlwyRR/vt8EfrCqvh94CbAzycu7jdS6twD3dh1iA/3jqnpJVU3VH4pD/ApwU1W9EPh+pvRzrqr7ms/0JcBLgYeBD3Wbqj1JNgM/BZxZVS+if4GekRffmURJXgS8CdhB/2f4NUle0G2qdXcNS/+muAz4SFVtAz7SrE+Da1ja17uAfwZ8dMPTtO8alvb3VuBFVfVi4E+At290qBZdw9L+vreqXtx8P38YuHzxiybUNSxTCyTZCvwT4PMbHUjtm6oCnv4v1oNVdaiqHgWuB3Z1nKk1VfVR+lcrnAlV9YWq+njz+P/RLwI2d5uqHdX3jWZ1oVmm9oIVSbYA/xS4qussWl9Jng2cQ/8qq1TVo1X1l52G2hjnAX9WVZ/rOkjL5oETmnvXPgN4oOM8bfle4I6qeriqjgL/m36xNzWG/E2xC7i2eXwt8NqNzNSW5fpaVfdW1X0dRWrVkP7e0vwsA9xO//7UU2FIf78+sHoiU/I31Qq1wC8D/4Yp6ae+1bQV8JuB+wfWDzOlBd6sS3IqcDpwR8dRWtNMKb8TeBC4taqmtq/Af6L/i+aJjnNslAJuSfLHSXZ3HaZlpwFfAv5Hc4rEVUlO7DrUBrgQuK7rEG2qqiPAf6Q/wvMF4GtVdUu3qVpzF3B2kucmeQbwamBrx5k2wslV9YXm8V8AJ3cZRq35l8DvdR2ibUl+Icn9wI8zPSPwSyTZBRypqk92nUXtmLYCXjMgyTOB3wLeuuiI6lSpqsebqV5bgB3NFM6pk+Q1wINV9cddZ9lA/6iqzqB/us8lSc7pOlCL5oEzgF+tqtOBv2J6puEuK8km4ALgN7vO0qbmfOhd9A/SnAKcmOQnuk3Vjqq6F3gPcAtwE3An8HiXmTZa9W9b5GjelEnyDvqnKH6g6yxtq6p3VNVW+n29tOs8bWgOMP4sU3yAQtNXwB/hW4+Ib2m2aUokWaBfvH+gqn676zwboZlufBvTe72Ds4ALknyW/mkvP5jkN7qN1K5m5JKqepD+OdI7uk3UqsPA4YEZJDfSL+in2fnAx6vqi10HadkrgD+vqi9V1WPAbwP/sONMramq91XVS6vqHOCr9M8bnnZfTPI8gObfBzvOo3WU5A3Aa4Afr9m6r/QHgB/qOkRL/i79g6qfbP6u2gJ8PMl3dJpK62raCvj9wLYkpzUjIBcCezvOpHWSJPTPo723qn6p6zxtSvJtx64Im+QE4JXAZzoN1ZKqentVbamqU+n/P/v7VTWVo3gASU5M8qxjj+lfZGZq7yRRVX8B3J/ke5pN5wH3dBhpI1zElE+fb3weeHmSZzTfz+cxpRcoBEjy7c2/z6d//vsHu020IfYCr28evx74nQ6zaB0l2Un/1LULqurhrvO0Lcm2gdVdTO/fVJ+uqm+vqlObv6sOA2c0v4s1Jea7DrCequpokkuBm+lfDffqqrq741itSXIdcC5wUpLDwDur6n3dpmrVWcC/AD7dnBsO8LNVta+7SK15HnBtc2eFOeCGqpr626vNiJOBD/XrHeaBD1bVTd1Gat2bgQ80B1YPAW/sOE9rmoMyrwR+sussbauqO5LcCHyc/hTcTwB7uk3Vqt9K8lzgMeCSabsY43J/UwDvBm5IcjHwOeBHu0u4fob09SvAfwG+DfjdJHdW1au6S7l+hvT37cDTgFub30e3V9W/6izkOhrS31c3B5KfoP+zPLV9nfJaQEBma8aMJEmSJEmTadqm0EuSJEmSNJUs4CVJkiRJmgAW8JIkSZIkTQALeEmSJEmSJoAFvCRJkiRJE8ACXpIkSZKkCWABL0mSJEnSBLCAlyRJkiRpAvx/nR3JyvjFMzoAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 1440x144 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "plt.rcParams[\"figure.figsize\"] = (20, 2) # (w, h)\n",
    "\n",
    "sb.heatmap(grch37_ds.one_hot_encode(all_chr).T,cmap = sb.color_palette(\"rocket_r\", as_cmap=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "8efd5b92-1d8a-4904-bf4e-3ba3b35616fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'encoding_dict': {'A': 0, 'C': 1, 'T': 2, 'G': 3, 'N': 4, 'M': 4, 'R': 4},\n",
       " 'available_chromosomes': ['20', '21'],\n",
       " 'chr_lengths': {'20': 63025520, '21': 48129895},\n",
       " 'time created': 1665672883.459766,\n",
       " 'num_2_base_dict': {0: 'A', 1: 'C', 2: 'T', 3: 'G', 4: 'R'},\n",
       " 'num_reverse_compliment_dict': {0: 2, 2: 0, 1: 3, 3: 1, 4: 4}}"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grch37_ds.metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7070011e-234b-4417-aedd-acc70c51e584",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
